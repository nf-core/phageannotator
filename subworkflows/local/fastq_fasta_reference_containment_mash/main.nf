//
// Identify sequences contained in readset
//

include { MASH_SKETCH as MASH_SKETCH_ASSEMBLIES } from '../../../modules/nf-core/mash/sketch/main'      // TODO: Update nf-core module to remove optional -r argument
include { MASH_SKETCH as MASH_SKETCH_REFERENCES } from '../../../modules/nf-core/mash/sketch/main'      // TODO: Update nf-core module to remove optional -r argument
include { MASH_PASTE                            } from '../../../modules/local/mash/paste/main'         // TODO: Add module to nf-core
include { MASH_SCREEN                           } from '../../../modules/nf-core/mash/screen/main'      // TODO: Update nf-core module to add meta to sketch

workflow FASTQ_FASTA_REFERENCE_CONTAINMENT_MASH {
    take:
    fastq_gz                // [ [ meta.id ] , [ 1.fastq.gz, 2.fastq.gz ]  , reads (mandatory)
    assembly_fasta_gz       // [ [ meta.id ] , fasta.gz ]                  , assemblies (optional)
    reference_fasta_gz      // [ [ meta.id ] , fasta.gz ]                  , reference sequences (mandatory)
    reference_sketch_msh    // [ [ meta.id ] , fasta.msh ]                 , reference_sequences_sketch (optional)

    main:
    ch_versions = Channel.empty()

    //
    // MODULE: Create a sketch of assemblies
    //
    ch_assembly_sketch_msh = MASH_SKETCH_ASSEMBLIES ( assembly_fasta_gz ).mash
    ch_versions = ch_versions.mix(MASH_SKETCH_ASSEMBLIES.out.versions.first())

    // if provided, use reference sketch. If not, create one
    if ( reference_sketch_msh ) {
        ch_reference_sketch_msh = reference_sketch_msh
    } else {
        //
        // MODULE: Create sketch of reference sequences
        //
        ch_reference_sketch_msh = MASH_SKETCH_REFERENCES ( reference_fasta_gz ).mash
        ch_versions = ch_versions.mix(MASH_SKETCH_REFERENCES.out.versions.first())
    }

    //
    // MODULE: Combine assembly and reference sketches
    //
    ch_combined_sketch_msh = MASH_PASTE ( ch_assembly_sketch_msh, ch_reference_sketch_msh.collect() ).msh
    ch_versions = ch_versions.mix(MASH_PASTE.out.versions.first())

    // join reads and combined sketch by meta.id
    ch_mash_screen_input = fastq_gz.join( ch_combined_sketch_msh, by:0 )

    //
    // MODULE: Identify contained genomes
    //
    ch_mash_screen_tsv = MASH_SCREEN ( ch_mash_screen_input.map { [ it[0], it[1] ] }, ch_mash_screen_input.map { [ it[0], it[2] ] } ).screen
    ch_versions = ch_versions.mix(MASH_SCREEN.out.versions.first())

    emit:
    mash_screen_results = ch_mash_screen_tsv    // [ [ meta.id ] , fasta.gz ]   , concatenated assemblies and contained references
    versions            = ch_versions           // [ versions.yml ]
}
