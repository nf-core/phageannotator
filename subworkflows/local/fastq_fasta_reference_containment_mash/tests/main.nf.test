nextflow_workflow {

    name "Test subworkflow: FASTQ_FASTA_REFERENCE_CONTAINMENT_MASH"
    script "subworkflows/local/fastq_fasta_reference_containment_mash/main.nf"
    workflow "FASTQ_FASTA_REFERENCE_CONTAINMENT_MASH"

    // Dependencies
    tag "MASH_SKETCH"
    tag "MASH_PASTE"
    tag "MASH_SCREEN"

    test("fastq.gz & fasta.gz") {

        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz', checkIfExists:true),
                            file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz', checkIfExists: true)
                        ]
                    ],
                    [
                        [ id:'test2' ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test2_1.fastq.gz', checkIfExists:true),
                            file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test2_2.fastq.gz', checkIfExists: true)
                        ]
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/genome/genome.fna.gz', checkIfExists:true)\
                    ],
                    [
                        [ id:'test2' ],
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/genome/genome.fna.gz', checkIfExists:true)
                    ]
                )
                input[2] = Channel.of(
                    [
                        [ id:'reference_viruses' ],
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkIfExists:true)
                    ]
                )
                input[3] = null
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
