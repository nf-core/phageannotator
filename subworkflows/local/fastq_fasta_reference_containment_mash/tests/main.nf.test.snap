{
    "fastq.gz & fasta.gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test2"
                        },
                        "test2.screen:md5,a50700d95a3bc0c9913997f142817bbd"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test.screen:md5,9cd27acafbf63f1fa92e26c3e83bfa3e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,418b54f7ff265f4ff933e51868b82d47",
                    "versions.yml:md5,481065b75b92012f105cefce45278e0b",
                    "versions.yml:md5,c1d295c206d08ab24a095537745ee679",
                    "versions.yml:md5,e516cb0998c459ba772f1b5a55696f2c"
                ],
                "mash_screen_results": [
                    [
                        {
                            "id": "test2"
                        },
                        "test2.screen:md5,a50700d95a3bc0c9913997f142817bbd"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test.screen:md5,9cd27acafbf63f1fa92e26c3e83bfa3e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,418b54f7ff265f4ff933e51868b82d47",
                    "versions.yml:md5,481065b75b92012f105cefce45278e0b",
                    "versions.yml:md5,c1d295c206d08ab24a095537745ee679",
                    "versions.yml:md5,e516cb0998c459ba772f1b5a55696f2c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-28T16:49:00.33025338"
    }
}
